Support Scala 2.13 in mlflow-spark

B-Step62 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/B-Step62/mlflow/pull/10505?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10505/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10505
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Update [mlflow-spark](https://github.com/databricks/mlflow-private/tree/6f0f0292278bc885b2b870f47500dd4fc46bf233/mlflow/java/spark) from Scala `2.12.6` to `2.13.10` A few notable things:
- Cannot update parent POM as mleap doesn't support scala 2.13 yet ([mleap/issues/811](https://github.com/combust/mleap/issues/811)) and mlflow-scoring depends on it.
- Updated Spark to 3.5.0 upon which DBR 14.x is built ([ref](https://docs.databricks.com/en/release-notes/runtime/index.html))
  - This requires a small change in autologging logic, as `DataSourceV2Relation` was removed from logical plan since 3.0.0 stable release. It was introduced in 3.0.0-preview (= current version we are using), but pulled back due to some issues. Hence the new logic fetches data source information from `LogicalRelation` node instead. The new logic is indeed just same as what we have for "REPL-ID aware" environment such as Databricks, so I assume the change is safe.
- While Spark 3.5 supports scala 2.13, we still need use scala 2.12 when using autologging from Python, as Pyspark is running on scala 2.12 anyway (before 4.0.0). As a workaround, we have to serve two jars - `mlflow-spark_2.12` and `mlflow-spark-2.13` meanwhile.
- Enabled cross-version tests (previously only enabled for 3.0.0-preview). 
  - Initially the tests were failed with Pyspark >= 3.2.0. Turned out it was due to "pin thread" mode that was defaulted since 3.2.0. As a workaround, setting env variable `PYSPARK_PIN_THREAD` to 'false'. See related code change for more details.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [x] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

B-Step62 committed:
Update mlflow-spark to use scala 2.13

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Workaround for DataSourceV2Relation issue

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Fix expired links

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Further bump for scala/spark

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented on mlflow/java/pom.xml:
```diff
@@ -159,6 +144,11 @@
         <version>${slf4j.version}</version>
       </dependency>
+      <dependency>
+        <groupId>org.slf4j</groupId>
```
ref: https://www.slf4j.org/codes.html#StaticLoggerBinder

B-Step62 commented on mlflow/java/spark/src/test/scala/org/mlflow/spark/autologging/SparkAutologgingIntegrationSuite.scala:
```diff
@@ -205,13 +205,15 @@ class SparkAutologgingSuite extends FunSuite with Matchers with BeforeAndAfterAl
     df.collect()
     Thread.sleep(1000)
-    verify(subscriber, times(1)).notify(any(), any(), any())
+    verify(subscriber, times(1)).notify(any(), any(), meq("text")) // Prefetch on load()
```
Somehow this broad match started to raise false-assert after version bump, even though the subscriber gets exact same calls and args. I guess it's just random ordering flakiness but can dig into more if this looks a concern.

B-Step62 committed:
Support scala 2.13 as separate maven profile, as pyspark still needs 2.12 jar.

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

WeichenXu123 commented on mlflow/java/spark/pom.xml:
```diff
@@ -7,8 +7,11 @@
+    <!-- Overriding Scala version to 2.13 for mlflow-spark.
+      We can't bump it up globally because mleap doesn't support scala 2.13 yet, which is dependency of mlflow-scoring.
+      Ref: https://github.com/combust/mleap/issues/811 -->
+    <scala.version>2.13.10</scala.version>
```
I think we should still support scala 2.12, spark 3.5 still supports scala 2.12.

We can support to build mlflow-spark with either scala-2.12 or scala 2.13, like this project:

https://github.com/dmlc/xgboost/blob/3f4e22015a6796a0807650d7b2f97b108f8760df/jvm-packages/xgboost4j-spark/pom.xml#L12

and we can release 2 jars , one for scala 2.12, one for scala 2.13, like:
https://mvnrepository.com/artifact/ml.dmlc/xgboost4j-spark

WDYT ?

B-Step62 replied:
Yeah I faced exact issue when calling from Pyspark, and was working exactly same approach:) 
Will push another revision soon!

WeichenXu123 commented on mlflow/java/spark/pom.xml:
```diff
@@ -1,13 +1,17 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
   <modelVersion>4.0.0</modelVersion>
-  <artifactId>mlflow-spark</artifactId>
+  <artifactId>mlflow-spark-${scala.compat.version}</artifactId>
```
According to convention,
we should use format like `mlflow-spark_{scala.compat.version}`

WeichenXu123 commented on mlflow/java/spark/pom.xml:
```diff
@@ -137,9 +170,4 @@
   </build>
-  <profiles>
-    <profile>
-      <id>do-sign</id>
```
Can we remove this ?

B-Step62 replied:
Yup, but not super sure what it does?. I saw [your PR](https://github.com/mlflow/mlflow/pull/4181) to adding this profile, but would removing it change the process of publishing artifacts to the maven central?

WeichenXu123 replied:
It is for signing jars when publishing, we define `do-sign` profile in parent pom, so when running mvn with this profile enabled, the generated jars will be signed. But I think sub-project should inherit the `do-sign` profile for supporting signing.

WeichenXu123 commented on mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/DatasourceAttributeExtractor.scala:
```diff
@@ -49,10 +49,20 @@ private[autologging] trait DatasourceAttributeExtractorBase {
       leafNode match {
+        // DataSourceV2Relation was removed in Spark 3.0.0 stable release due to some issue and 
+        // still not present in Spark 3.2.0. While we are not sure whether it will be back in
+        // the future, we still keep this code here to support previous versions.
         case relation: DataSourceV2Relation =>
           getSparkTableInfoFromTable(relation.table)
-        case other =>
-          None
+        // This is the case for Spark 3.x except for 3.0.0-preview
+        case LogicalRelation(HadoopFsRelation(index, _, _, _, fileFormat, _), _, _, _) =>
+          val path: String = index.rootPaths.headOption.map(_.toString).getOrElse("unknown")
+          val formatOpt = fileFormat match {
+            case format: DataSourceRegister => Option(format.shortName)
+            case _ => None
+          }
+          Option(SparkTableInfo(path, None, formatOpt))
+        case _ => None
```
CC @cloud-fan to take a look.

WeichenXu123 commented on mlflow/ml-package-versions.yml:
```diff
@@ -351,17 +351,12 @@ spark:
-      fi
+      # Build Java package
+      pushd mlflow/java/spark
+      mvn package -DskipTests -q
```
I think we need to use `mvn install ...` if we want to run test in following script ?

B-Step62 committed:
Set PYSPARK_PIN_THREAD to false for workarounding Py4j issue

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Fix link

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

cloud-fan commented on mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/DatasourceAttributeExtractor.scala:
```diff
@@ -49,10 +49,20 @@ private[autologging] trait DatasourceAttributeExtractorBase {
-          None
+        // This is the case for Spark 3.x except for 3.0.0-preview
+        case LogicalRelation(HadoopFsRelation(index, _, _, _, fileFormat, _), _, _, _) =>
+          val path: String = index.rootPaths.headOption.map(_.toString).getOrElse("unknown")
```
Spark can read from multiple paths in one shot. Do we want to display all of them or only the first path?

B-Step62 committed:
Fix unit tests

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Fix hidden inner-test dependency

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Weaken PYSPARK_PIN_THREAD validation from exception to warning

Raising exception affects many CI tests even though they don't
really need datasource autologging. We don't really need to
throw exception as it would raise error when the actual event
is sent from JVM.

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Update doc

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Support pyspark dev version (4.x)

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Address comments

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented on mlflow/_spark_autologging.py:
```diff
@@ -235,7 +243,8 @@ class Java:
     """
     Context provider used when there's no active run. Accumulates datasource read information,
-    then logs that information to the next-created run & clears the accumulated information.
+    then logs that information to the next-created run. Note that this doesn't clear the accumlated
```
fyi: not changing behavior, just updated to the latest behavior (#4086)

B-Step62 commented on tests/spark/autologging/conftest.py:
```diff
@@ -2,34 +2,33 @@
 
 
-@pytest.fixture(scope="module")
+@pytest.fixture
```
Changing scope to function so as not to share same file paths across test, which might hide some failure.
For example, SparkAutologgingContext accumulates events to global variable `_table_info`, so those events can be carried over to the other test cases, unless cleaned explicitly (like I added in tear_down fixture). However, the unit tests have been passing so far because the file path in the event was same.

WeichenXu123 commented on mlflow/java/spark/pom.xml:
```diff
@@ -1,13 +1,17 @@
+      and raises an exception due to imcomptibility if mlflow-spark jar is built with 2.13.
+      This is temporary fix until (Py)Spark 4 release that will use scala 2.13 by default. -->
+    <scala.version>2.12.18</scala.version>
     <scala.compat.version>2.12</scala.compat.version>
```
A reminder on follow-up task: we need to update MLflow release pipeline, to update stage of java releasing, so that it can release both scala 2.12 and 2.13 version

WeichenXu123 reviewed (approved):
LGTM, but let's waiting for @cloud-fan 's approval

cloud-fan reviewed (approved):
Spark part LGTM

harupy commented on mlflow/spark/__init__.py:
```diff
@@ -1106,6 +1107,20 @@ def autolog(disable=False, silent=False):  # pylint: disable=unused-argument
+    # However, this causes an issue in spark autologging as our event listener thread may be
+    # terminated before receiving the datasource event.
+    # Hence, we have to disable it and decouple the thread management between Python and JVM.
+    if pyspark_version >= "3.2" and os.environ.get("PYSPARK_PIN_THREAD", "").lower() != "false":
```
Comparing versions as a string can lead to errors. Can we use `packaging`?

B-Step62 replied:
Sure! Fixing now.

B-Step62 committed:
Avoid use string comparison for version check

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
lint

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented on tests/spark/autologging/conftest.py:
```diff
@@ -49,3 +48,16 @@ def format_to_file_path(spark_session):
         yield res
+
+
+@pytest.fixture(autouse=True)
+def tear_down():
+    yield
+
+    # Clear cached table infos. When the datasource event from Spark arrives but there is no
+    # active run (e.g. the even comes with some delay), MLflow keep them in memory and logs them to
+    # the next **and any successive active run** (ref: PR #4086).
+    # However, this behavior is not desirable during tests, as we don't want any tests to be
+    # affected by the previous test. Hence, this fixture is executed on every test function
+    # to clear the accumulated table infos stored in the global context.
+    clear_table_infos()
```
Can we apply `@pytest.fixture(autouse=True)` on `clear_table_infos`?

B-Step62 replied:
Do you mean putting pytest annotation to the original def in `_spark_autologging.py`?

harupy replied:
I see, `clear_table_infos` is imported from other module. Never mind.

harupy commented on tests/spark/autologging/utils.py:
```diff
@@ -26,13 +26,23 @@ def _get_expected_table_info_row(path, data_format, version=None):
     assert _SPARK_TABLE_INFO_TAG_NAME not in run.data.tags
 
 
+def _disable_pin_thread():
+    # PYSPARK_PIN_THREAD is set to true by default since Pyspark 3.2.0, which causes
+    # issues with Py4J callbacks, so we ask users to set it to false.
+    # We have to set this before creating the SparkSession.
+    os.environ["PYSPARK_PIN_THREAD"] = "false"
```
Can we use the `monkeypatch` fixture instead of mutating `os.environ`?

harupy commented on mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala:
```diff
@@ -149,12 +149,17 @@ private[autologging] trait MlflowAutologEventPublisherImpl {
   private val checkpointFilePattern = ".*\\d+\\.checkpoint(\\.\\d+\\.\\d+)?\\.parquet$".r.pattern
   private def isCheckpointFile(path: String): Boolean = checkpointFilePattern.matcher(path).matches()
 
+  private def shouldSkipPublish(path: String, format: Option[String]): Boolean = {
+    // 1. Spark first loads head of the data as unknown "text" to infer the schema, which we don't want to log
+    // 2. Checkpoint files don't provide useful information, so we filter them out
+    (format.isEmpty || format.get == "text") || isCheckpointFile(path)
+  }
+
```
/ai Is `format.forall(f => f == "text")` equivalent to `(format.isEmpty || format.get == "text")`?

harupy reviewed (approved):
Left a few more comments, otherwise LGTM!

B-Step62 committed:
Use monkeypatch for env tweak

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

xinzhao-db commented on mlflow/java/spark/pom.xml:
```diff
@@ -29,17 +44,32 @@
-      <version>3.0.0-preview</version>
+      <version>3.5.0</version>
+    </dependency>
+    <dependency>
```
why do we add these dependencies? It seems only mlflow/java/scoring are using them

B-Step62 replied:
mlflow/java/spark uses them: https://github.com/mlflow/mlflow/blob/master/mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/ReplAwareSparkDataSourceListener.scala#L3-L6

xinzhao-db commented on mlflow/java/spark/pom.xml:
```diff
@@ -59,10 +89,14 @@
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
     </dependency>
+    <dependency>
```
and this one?
